dali/operators/util/get_property.cc

```diff
@@ -0,0 +1,162 @@
+would be to )code")
+    .NumInput(1)
+    .NumOutput(1)
+    .AddArg("key",
+            R"code(Specifies, which property is requested. The following properties are supported:
+
+* ``"source_info"``: Returned type: byte-array.
+                     Information about the origin of the sample. The actual origin may differ according
+                     to the source of the data. For example, when the Tensor is read via :meth:`fn.readers.file`,
+                     the ``source_info`` property will contain full path of that file. When the Tensor
+                     is read by :meth:`fn.readers.webdataset`, the property will contain full path of the tar
+                     archive and the index of the component in that archive. Lastly, when the Tensor
+                     is loaded via :meth:`fn.external_source`, the ``source_info`` will be empty.
+* ``"layout"``: Returned type: byte-array
+                Data layout in the given Tensor.
+)code",
+            DALI_STRING);
+
+
+DALI_REGISTER_OPERATOR(GetProperty, GetProperty<CPUBackend>, CPU)
+DALI_REGISTER_OPERATOR(GetProperty, GetProperty<GPUBackend>, GPU)
+
+namespace detail {
+namespace {
+
+template <typename Backend, typename BatchContainer = batch_container_t<Backend>>
+struct SourceInfo : public Property<Backend> {
+  TensorListShape<> GetShape(const BatchContainer& input) override {
+    TensorListShape<> ret{static_cast<int>(input.num_samples()), 1};
+    for (int i = 0; i < ret.size(); i++) {
+      ret.set_tensor_shape(i, {static_cast<int>(GetSourceInfo(input, i).length())});
+    }
+    return ret;
+  }
+
+  DALIDataType GetType(const BatchContainer&) override {
+    return DALI_UINT8;
+  }
+
+  void FillOutput(workspace_t<Backend>& ws) override {
+    const auto& input = ws.template Input<Backend>(0);
+    auto& output = ws.template Output<Backend>(0);
+    for (size_t sample_id = 0; sample_id < input.num_samples(); sample_id++) {
+      auto si = GetSourceInfo(input, sample_id);
+      output[sample_id].Copy(make_cspan((const uint8_t*)si.c_str(), si.length()), nullptr);
+    }
+  }
+
+ private:
+  const std::string& GetSourceInfo(const BatchContainer& input, size_t idx) {
+    return input[idx].GetMeta().GetSourceInfo();
+  }
+};
+
+template <>
+const std::string& SourceInfo<GPUBackend>::GetSourceInfo(const TensorList<GPUBackend>& input,
+                                                         size_t idx) {
+  return input.GetMeta(static_cast<int>(idx)).GetSourceInfo();
+}
+
+template <>
+void SourceInfo<GPUBackend>::FillOutput(workspace_t<GPUBackend>& ws) {
+  const auto& input = ws.template Input<GPUBackend>(0);
+  auto& output = ws.template Output<GPUBackend>(0);
+  for (size_t sample_id = 0; sample_id < input.num_samples(); sample_id++) {
+    auto si = GetSourceInfo(input, sample_id);
+    auto output_ptr = output.raw_mutable_tensor(static_cast<int>(sample_id));
+    assert(ws.has_stream());  // GPU operator must have CUDA stream assigned
```

Comment:
Do we need this? I think we can consider this as granted.

Reply:
Removed